include/revng/AutoEnforcer/AutoEnforcer.h

```diff
@@ -0,0 +1,183 @@
+  }
+
+  template<typename BackingContainerType>
+  BackingContainerType &getOrCreateFinalContainer(llvm::StringRef Name) {
+    return Pipeline.getOrCreateFinalContainer<BackingContainerType>(Name);
+  }
+
+  llvm::Error
+  run(const BackingContainersStatus &Targets, llvm::raw_ostream *OS = nullptr) {
+    return Pipeline.run(*AEContext, Targets, OS);
+  }
+
+  template<typename BackingContainerType, typename... Args>
+  void registerContainerFactory(llvm::StringRef Name, Args &&...Arguments) {
+    revng_assert(not CommittedRegistry);
+    using BCT = BackingContainerType;
+    Registry.registerContainerFactory<BCT>(Name,
+                                           std::forward<Args>(Arguments)...);
+  }
+
+  void
+  registerContainerFactory(llvm::StringRef Name,
+                           std::unique_ptr<BackingContainerFactory> Entry) {
+    Registry.registerContainerFactory(Name, std::move(Entry));
+  }
+
+  template<typename BackingContainer>
+  void registerDefaultConstructibleFactory(llvm::StringRef Name) {
+    revng_assert(not CommittedRegistry);
+    Registry.registerDefaultConstructibleFactory<BackingContainer>(Name);
+  }
+
+  iterator begin() { return Pipeline.begin(); }
+  iterator end() { return Pipeline.end(); }
+
+  const_iterator begin() const { return Pipeline.begin(); }
+  const_iterator end() const { return Pipeline.end(); }
+
+  template<typename OStream>
+  void dump(OStream &OS, size_t Indents = 0) const {
+    Pipeline.dump(OS, Indents);
+  }
+  void dump() const debug_function { dump(dbg); }
+  Step &operator[](size_t Index) { return Pipeline[Index]; }
+  const Step &operator[](size_t Index) const { return Pipeline[Index]; }
+
+  const Step &back() const { return Pipeline.back(); }
+  Step &back() { return Pipeline.back(); }
+
+  llvm::Expected<InvalidationMap>
+  getInvalidations(const AutoEnforcerTarget &Target) const;
+
+  llvm::Error invalidate(const AutoEnforcerTarget &Target);
+  llvm::Error invalidate(const InvalidationEventBase &Event);
+
+  llvm::Error store(llvm::StringRef DirPath) const {
+    if (auto Error = Pipeline.store(DirPath); !!Error)
+      return Error;
+    return AEContext->save(DirPath);
+  }
+
+  llvm::Error load(llvm::StringRef DirPath) {
+    if (auto Error = AEContext->load(DirPath); !!Error)
+      return Error;
+    return Pipeline.load(DirPath);
+  }
+
+  llvm::Expected<const BackingContainerBase *>
+  safeGetContainer(llvm::StringRef StepName,
+                   llvm::StringRef ContainerName) const;
+  llvm::Expected<BackingContainerBase *>
+  safeGetOrCreateContainer(llvm::StringRef StepName,
+                           llvm::StringRef ContainerName);
+
+  TargetsList deduceInvalidations(const InvalidationEventBase &Event) {
+    return AEContext->getKindRegistry().deduceInvalidations(Event);
+  }
+
+  const KindsRegisty &getKindRegistry() const {
+    return AEContext->getKindRegistry();
+  }
+
+private:
+  Context *AEContext;
```

Comment:
`AE`? If you don't want to repeat `Context`, our convention is `TheContext`. Not ideal, but yeah.